README.md
calmsize
==============

### NOTICE

I forked the repo from source file downloaded from pip `hurry.filesize`.
That simple library is very useful when debugging memory issues, but 
unfortunately out-of-maintainance since 2009, so I decided to fork
the great work of Martijn Faassen, Startifact.

### Intro

calmsize a simple Python library that can take a number of bytes and
returns a human-readable string with the size in it, in kilobytes (K),
megabytes (M), etc.

The default system it uses is "traditional", where multipliers of 1024
increase the unit size::

```ipython
>>> from calmsize import size, ByteSize
>>> size(1024)
'1K'
>>> size(-1024)
'-1K'
>>> '{:.2f}'.format(size(-1024))
'-1.00K'
```

An alternative, slightly more verbose system::

```ipython
>>> from calmsize import alternative
>>> size(1, system=alternative)
'1 byte'
>>> size(10, system=alternative)
'10 bytes'
>>> size(1024, system=alternative)
'1 KB'
```

A verbose system::

```ipython
>>> from calmsize import verbose
>>> size(10, system=verbose)
'10 bytes'
>>> size(1024, system=verbose)
'1 kilobyte'
>>> size(2000, system=verbose)
'1 kilobyte'
>>> size(3000, system=verbose)
'2 kilobytes'
>>> size(1024 * 1024, system=verbose)
'1 megabyte'
>>> size(1024 * 1024 * 3, system=verbose)
'3 megabytes'
```

You can also use the SI system, where multipliers of 1000 increase the unit
size::

```ipython
>>> from calmsize import si
>>> size(1000, system=si)
'1K'
```


CHANGES.md
Changes
=======

0.1.2 (2019-05-27)
----------------
* Fix pip install fail of python2.7


0.1.1 (2019-05-25)
----------------
* Fix pip install filenotfound error


0.1.0 (2019-05-24)
----------------

* Add support for negative input
* Add support for format the leading number
* Add ByteSize as a new class
* Add support for comparing between two Bytesize, ByteSize and numbers


0.0.1 (2019-05-24)
----------------

* Forked from hurry.filesize


test/test_basic.py
import pytest
from calmsize import size, ByteSize

def test_basics():
    assert size(10) == '10B'
    assert size(1024) == '1K'
    assert size(1024 * 1024 * 10) == '10M'

def test_neg():
    assert size(-10) == '-10B'
    assert size(-1024) == '-1K'
    assert size(-1024 * 1024 * 10) == '-10M'

def test_format():
    assert '{:.2f}'.format(size(12345678)) == '11.77M'

def test_round():
    assert size(12345678) == '12M'

def test_compare():
    assert size(-10240) == '-10K'
    with pytest.raises(NotImplementedError) as e:
        size(-10241) < '-10K'
    with pytest.raises(NotImplementedError) as e:
        size(-10241) > '-10K'

    assert size(-10241) < size(-10240)
    assert size(-10241) > -10240
    assert size(-10241) > -10240.888



calmsize/calmsize.py
# encoding=utf-8
# Modified by Kaiyu Shi
# Date:2019-05-24

traditional = [
    (1024 ** 5, 'P'),
    (1024 ** 4, 'T'),
    (1024 ** 3, 'G'),
    (1024 ** 2, 'M'),
    (1024 ** 1, 'K'),
    (1024 ** 0, 'B'),
    ]

alternative = [
    (1024 ** 5, ' PB'),
    (1024 ** 4, ' TB'),
    (1024 ** 3, ' GB'),
    (1024 ** 2, ' MB'),
    (1024 ** 1, ' KB'),
    (1024 ** 0, (' byte', ' bytes')),
    ]

verbose = [
    (1024 ** 5, (' petabyte', ' petabytes')),
    (1024 ** 4, (' terabyte', ' terabytes')),
    (1024 ** 3, (' gigabyte', ' gigabytes')),
    (1024 ** 2, (' megabyte', ' megabytes')),
    (1024 ** 1, (' kilobyte', ' kilobytes')),
    (1024 ** 0, (' byte', ' bytes')),
    ]

iec = [
    (1024 ** 5, 'Pi'),
    (1024 ** 4, 'Ti'),
    (1024 ** 3, 'Gi'),
    (1024 ** 2, 'Mi'),
    (1024 ** 1, 'Ki'),
    (1024 ** 0, ''),
    ]

si = [
    (1000 ** 5, 'P'),
    (1000 ** 4, 'T'),
    (1000 ** 3, 'G'),
    (1000 ** 2, 'M'),
    (1000 ** 1, 'K'),
    (1000 ** 0, 'B'),
    ]


class ByteSize():
    def __init__(self, num_bytes, system=traditional):
        self.num_bytes = num_bytes
        self.system = system
        self.amount = num_bytes
        self.unit = self.system[-1]  # lowest is pure Bytes
        self.find_largest_unit()

    def _find_largest_unit_pos(self, num_bytes):
        """Find the proper unit and corresponding amount

        This implementation only works for positive number
        """
        for factor, unit in self.system:
            if num_bytes >= factor:
                break
        self.amount = num_bytes / factor

        # singular and plural for a tuple
        if isinstance(unit, tuple):
            singular, multiple = unit
            if self.amount == 1:
                unit = singular
            else:
                unit = multiple
        self.unit = unit

    def find_largest_unit(self):
        num_bytes = self.num_bytes
        pos_bytes = abs(num_bytes)
        sign = int(num_bytes >= 0) * 2 - 1  # sign function
        self._find_largest_unit_pos(pos_bytes)
        self.amount *= sign

    def __str__(self):
        return str(int(round(self.amount))) + self.unit

    def __format__(self, formatstr):
        if formatstr:
            return self.amount.__format__(formatstr) + self.unit
        else:
            return str(self)

    def __repr__(self):
        return str(self)

    def __eq__(self, other):
        if isinstance(other, str):
            return str(self) == other
        elif isinstance(other, ByteSize):
            return self.amount == other.amount
        else:
            return type(other)(self.amount) == other

    def __lt__(self, other):
        if isinstance(other, str):
            raise NotImplementedError(
                'Comparison between string and ByteSize not supported yet'
            )
        elif isinstance(other, ByteSize):
            return self.amount < other.amount
        else:
            return type(other)(self.amount) < other

    def __gt__(self, other):
        if isinstance(other, str):
            raise NotImplementedError(
                'Comparison between string and ByteSize not supported yet'
            )
        elif isinstance(other, ByteSize):
            return self.amount > other.amount
        else:
            return type(other)(self.amount) > other


def size(bytes, system=traditional):
    """Human-readable file size.

    Using the traditional system, where a factor of 1024 is used::

    >>> size(10)
    '10B'
    >>> size(100)
    '100B'
    >>> size(2000000)
    '1M'

    Using the SI system, with a factor 1000::

    >>> size(10, system=si)
    '10B'
    >>> size(100, system=si)
    '100B'
    >>> size(1000, system=si)
    '1K'
    >>> size(2000000, system=si)
    '2M'

    """
    byte_size = ByteSize(bytes, system)
    return byte_size
